File: app2/src/main/java/ru/neverlands/anclient/postfilter/MineAjaxPhp.java
package ru.neverlands.anclient.postfilter;

import android.text.TextUtils;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import ru.neverlands.anclient.manager.AutoFunctionsManager;
import ru.neverlands.anclient.manager.AutoMineManager;
import ru.neverlands.anclient.utils.AppLog;
import ru.neverlands.anclient.utils.AppVars;
import ru.neverlands.anclient.utils.ChatStats;
import ru.neverlands.anclient.utils.Russian;

/** Post-filter for gameplay/ajax/mine_ajax.php Auto-Mine responses. */
public final class MineAjaxPhp {
    private static final String TAG = "MineAjaxPhp";
    private static final int AUTO_DIG_EXTRA_DELAY_SECONDS = 2;
    private static final Pattern DIGG_REPORT_PATTERN = Pattern.compile(
            "(?:^|\\^)DIGG@([^@\\r\\n]+)@(\\d+)",
            Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE);
    private static final Pattern RESOURCE_ENTRY_PATTERN = Pattern.compile(
            "([^,.:]+?)\\s*\\((\\d+(?:[\\.,]\\d+)?)\\)",
            Pattern.UNICODE_CASE);

    private MineAjaxPhp() {
    }

    public static byte[] process(String address, byte[] array) {
        if (array == null || array.length == 0) {
            return array;
        }
        String html = Russian.getString(array);
        if (TextUtils.isEmpty(html)) {
            return array;
        }
        if (AppVars.getContext() == null) {
            return array;
        }
        AutoMineManager manager = AutoMineManager.getInstance(AppVars.getContext());
        boolean autoMineEnabled = isAutoMineEnabled();
        boolean pendingRoute = manager.hasPendingMineRoute();
        boolean torchRequired = html.toLowerCase(Locale.ROOT).contains("вам нужен факел");
        boolean tooTired = MapAjax.containsTooTiredMessage(html);
        if (!autoMineEnabled && !pendingRoute && !torchRequired && !tooTired) {
            return array;
        }
        if (tooTired) {
            manager.onMineRouteTooTiredFromServer(address, html);
        } else {
            manager.onMineAjaxResponse(address, html);
        }
        MineDiggReport diggReport = parseMineDiggReport(address, html);
        boolean newDiggReport = true;
        if (diggReport != null) {
            newDiggReport = manager.onMineDiggReport(diggReport.eventKey, diggReport.serverDelaySeconds);
            if (newDiggReport && !diggReport.deltaByResourceKg.isEmpty()) {
                ChatStats.addResourceDeltaKg(diggReport.deltaByResourceKg);
            }
            AppLog.i(AutoMineManager.TRACE_CHAIN, TAG,
                    "digg report parsed: resources=" + diggReport.resources.size()
                            + ", statsUpdated=" + (newDiggReport && !diggReport.deltaByResourceKg.isEmpty())
                            + ", serverDelaySec=" + diggReport.serverDelaySeconds
                            + ", effectiveDelaySec=" + diggReport.effectiveDelaySeconds);
        }
        if (autoMineEnabled) {
            publishMineChatReportIfNeeded(manager, address, html, diggReport, newDiggReport);
        }
        return array;
    }

    private static void publishMineChatReportIfNeeded(AutoMineManager manager,
                                                      String address,
                                                      String html,
                                                      MineDiggReport diggReport,
                                                      boolean newDiggReport) {
        if (manager == null || !manager.isChatReportEnabled()) {
            return;
        }
        String text = html == null ? "" : html;
        String lower = text.toLowerCase(Locale.ROOT);
        String message = "";
        if (diggReport != null) {
            if (!newDiggReport) {
                return;
            }
            message = buildDiggChatMessage(diggReport);
        } else if (lower.contains("обнаружены ресурсы")) {
            message = compactResourceText(text, lower, "обнаружены ресурсы");
        } else if (lower.contains("добыты ресурсы")) {
            message = compactResourceText(text, lower, "добыты ресурсы");
        } else if (lower.contains("вы не нашли ни одного ресурса")) {
            message = "Вы не нашли ни одного ресурса";
        }
        if (message.isEmpty()) {
            return;
        }
        String chatHtml = MainPhp.buildServerChatTimeHtmlExternal()
                + "<font color=#333399><b>[auto_mine/mine_ajax]</b></font> "
                + escapeHtml(message);
        MainPhp.sendInventoryChatMessage(chatHtml);
        AppLog.i(AutoMineManager.TRACE_CHAIN, TAG,
                "chat report posted: address=" + safe(address) + ", message=" + message);
    }

    private static MineDiggReport parseMineDiggReport(String address, String html) {
        if (TextUtils.isEmpty(html)) {
            return null;
        }
        Matcher matcher = DIGG_REPORT_PATTERN.matcher(html);
        if (!matcher.find()) {
            return null;
        }
        String reportText = compactText(matcher.group(1));
        int serverDelaySeconds = parsePositiveInt(matcher.group(2));
        List<MineResourceEntry> resources = parseResources(reportText);
        Map<String, Double> deltaByResourceKg = new LinkedHashMap<>();
        for (MineResourceEntry entry : resources) {
            Double existing = deltaByResourceKg.get(entry.name);
            deltaByResourceKg.put(entry.name, (existing == null ? 0d : existing) + entry.kilograms);
        }
        return new MineDiggReport(
                safe(address) + "|" + reportText + "|" + serverDelaySeconds,
                reportText,
                serverDelaySeconds,
                resources,
                deltaByResourceKg);
    }

    private static List<MineResourceEntry> parseResources(String reportText) {
        List<MineResourceEntry> result = new ArrayList<>();
        if (TextUtils.isEmpty(reportText)) {
            return result;
        }
        String resourcePart = reportText;
        int colon = resourcePart.indexOf(':');
        if (colon >= 0 && colon + 1 < resourcePart.length()) {
            resourcePart = resourcePart.substring(colon + 1);
        }
        int digIndex = resourcePart.toLowerCase(Locale.ROOT).indexOf("добываем");
        if (digIndex >= 0) {
            resourcePart = resourcePart.substring(0, digIndex);
        }
        Matcher matcher = RESOURCE_ENTRY_PATTERN.matcher(resourcePart);
        while (matcher.find()) {
            String name = compactText(matcher.group(1));
            Double kilograms = parseDouble(matcher.group(2));
            if (!name.isEmpty() && kilograms != null && kilograms > 0d) {
                result.add(new MineResourceEntry(name, kilograms, formatKgForChat(kilograms)));
            }
        }
        return result;
    }

    private static String buildDiggChatMessage(MineDiggReport report) {
        StringBuilder builder = new StringBuilder();
        builder.append("Ресурсы: ");
        if (report.resources.isEmpty()) {
            builder.append("не распознаны");
        } else {
            for (int i = 0; i < report.resources.size(); i++) {
                MineResourceEntry entry = report.resources.get(i);
                if (i > 0) {
                    builder.append(", ");
                }
                builder.append(entry.name)
                        .append(" (")
                        .append(entry.amountText)
                        .append(" кг)");
            }
        }
        if (report.effectiveDelaySeconds > 0) {
            builder.append(". Задержка: ")
                    .append(report.effectiveDelaySeconds)
                    .append(" сек.");
        }
        return builder.toString();
    }

    private static String compactResourceText(String text, String lower, String marker) {
        int index = lower.indexOf(marker);
        if (index < 0) {
            return compactText(text);
        }
        return compactText(text.substring(index));
    }

    private static boolean isAutoMineEnabled() {
        try {
            return AppVars.getContext() != null
                    && AutoFunctionsManager.getInstance(AppVars.getContext()).isAutoMineEnabled();
        } catch (Exception e) {
            AppLog.w(AutoMineManager.TRACE_CHAIN, TAG, "AutoMine state read failed", e);
            return false;
        }
    }

    private static String compactText(String value) {
        if (value == null) {
            return "";
        }
        String result = value.replace('\u00A0', ' ')
                .replaceAll("<[^>]+>", " ")
                .replaceAll("\\s+", " ")
                .trim();
        if (result.length() > 400) {
            result = result.substring(0, 400) + "...";
        }
        return result;
    }

    private static int parsePositiveInt(String raw) {
        if (raw == null) {
            return 0;
        }
        try {
            return Math.max(0, Integer.parseInt(raw.trim()));
        } catch (NumberFormatException ignored) {
            return 0;
        }
    }

    private static Double parseDouble(String raw) {
        if (raw == null) {
            return null;
        }
        try {
            return Double.parseDouble(raw.trim().replace(',', '.'));
        } catch (NumberFormatException ignored) {
            return null;
        }
    }

    private static String formatKgForChat(double value) {
        String formatted = String.format(Locale.US, "%.2f", value);
        while (formatted.contains(".") && formatted.endsWith("0")) {
            formatted = formatted.substring(0, formatted.length() - 1);
        }
        if (formatted.endsWith(".")) {
            formatted = formatted.substring(0, formatted.length() - 1);
        }
        return formatted;
    }

    private static String escapeHtml(String value) {
        if (value == null) return "";
        return value.replace("&", "&amp;")
                .replace("<", "&lt;")
                .replace(">", "&gt;")
                .replace("\"", "&quot;")
                .replace("'", "&#39;");
    }

    private static String safe(String value) {
        return value == null ? "" : value.trim();
    }

    private static final class MineDiggReport {
        final String eventKey;
        final String reportText;
        final int serverDelaySeconds;
        final int effectiveDelaySeconds;
        final List<MineResourceEntry> resources;
        final Map<String, Double> deltaByResourceKg;

        MineDiggReport(String eventKey,
                       String reportText,
                       int serverDelaySeconds,
                       List<MineResourceEntry> resources,
                       Map<String, Double> deltaByResourceKg) {
            this.eventKey = eventKey;
            this.reportText = reportText;
            this.serverDelaySeconds = serverDelaySeconds;
            this.effectiveDelaySeconds = serverDelaySeconds > 0
                    ? serverDelaySeconds + AUTO_DIG_EXTRA_DELAY_SECONDS : 0;
            this.resources = resources;
            this.deltaByResourceKg = deltaByResourceKg;
        }
    }

    private static final class MineResourceEntry {
        final String name;
        final double kilograms;
        final String amountText;

        MineResourceEntry(String name, double kilograms, String amountText) {
            this.name = name;
            this.kilograms = kilograms;
            this.amountText = amountText;
        }
    }
}

File: ANClient/PostFilter/MineAjaxPhp.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;
using ANClient.ANForms;

namespace ANClient.PostFilter
{
    internal static partial class Filter
    {
        private static readonly Regex MineDiggReportRegex = new Regex(
            "(?:^|\\^)DIGG@([^@\\r\\n]+)@(\\d+)",
            RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

        private static readonly Regex MineResourceRegex = new Regex(
            "([^,.:]+?)\\s*\\((\\d+(?:[\\.,]\\d+)?)\\)",
            RegexOptions.CultureInvariant);

        private static byte[] MineAjaxPhp(string address, byte[] array)
        {
            if (array == null || array.Length == 0)
                return array;

            var html = AppVars.Codepage.GetString(array);
            if (string.IsNullOrEmpty(html))
                return array;

            var shouldHandle = AppVars.DoAutoMine || AutoMineRuntime.HasPendingMineRoute() ||
                               html.IndexOf("вам нужен факел", StringComparison.OrdinalIgnoreCase) >= 0 ||
                               html.IndexOf("вам нужна кирка", StringComparison.OrdinalIgnoreCase) >= 0;
            if (!shouldHandle)
                return array;

            AutoMineRuntime.OnMineAjaxResponse(address, html);
            var report = ParseMineDiggReport(address, html);
            var isNewReport = true;
            if (report != null)
            {
                isNewReport = AutoMineRuntime.OnMineDiggReport(report.EventKey, report.ServerDelaySeconds);
                var statsUpdated = false;
                if (isNewReport && report.DeltaByResourceKg.Count > 0 && AppVars.MainForm != null)
                {
                    AppVars.MainForm.UpdateMineResourceStatsSafe(report.DeltaByResourceKg);
                    statsUpdated = true;
                }

                AppLog.i(AutoMineRuntime.TraceChain, "MineAjaxPhp", "DIGG report parsed: resources=" + report.Resources.Count.ToString(CultureInfo.InvariantCulture) + ", statsUpdated=" + statsUpdated + ", serverDelay=" + report.ServerDelaySeconds.ToString(CultureInfo.InvariantCulture) + ", new=" + isNewReport);
            }

            if (AppVars.DoAutoMine && AppVars.Profile != null && AppVars.Profile.AutoMineChatReport)
            {
                PublishMineChatReport(address, html, report, isNewReport);
            }

            return array;
        }

        private static MineDiggReport ParseMineDiggReport(string address, string html)
        {
            if (string.IsNullOrEmpty(html))
                return null;

            var match = MineDiggReportRegex.Match(html);
            if (!match.Success)
                return null;

            var reportText = CompactText(match.Groups[1].Value);
            int serverDelay;
            if (!int.TryParse(match.Groups[2].Value, NumberStyles.Integer, CultureInfo.InvariantCulture, out serverDelay))
                serverDelay = 0;

            var resources = ParseMineResources(reportText);
            return new MineDiggReport((address ?? string.Empty) + "|" + reportText + "|" + serverDelay.ToString(CultureInfo.InvariantCulture), reportText, serverDelay, resources, BuildMineResourceDelta(resources));
        }

        private static Dictionary<string, double> BuildMineResourceDelta(List<MineResourceEntry> resources)
        {
            var result = new Dictionary<string, double>(StringComparer.OrdinalIgnoreCase);
            if (resources == null)
                return result;

            for (var i = 0; i < resources.Count; i++)
            {
                var entry = resources[i];
                if (entry == null || string.IsNullOrEmpty(entry.Name) || entry.Kg <= 0d)
                    continue;

                if (result.ContainsKey(entry.Name))
                    result[entry.Name] += entry.Kg;
                else
                    result.Add(entry.Name, entry.Kg);
            }

            return result;
        }

        private static List<MineResourceEntry> ParseMineResources(string reportText)
        {
            var result = new List<MineResourceEntry>();
            if (string.IsNullOrEmpty(reportText))
                return result;

            var resourcePart = reportText;
            var colon = resourcePart.IndexOf(':');
            if (colon >= 0 && colon + 1 < resourcePart.Length)
                resourcePart = resourcePart.Substring(colon + 1);

            var digIndex = resourcePart.IndexOf("Добываем", StringComparison.OrdinalIgnoreCase);
            if (digIndex >= 0)
                resourcePart = resourcePart.Substring(0, digIndex);

            var matcher = MineResourceRegex.Match(resourcePart);
            while (matcher.Success)
            {
                double kg;
                if (double.TryParse(matcher.Groups[2].Value.Replace(',', '.'), NumberStyles.Any, CultureInfo.InvariantCulture, out kg) && kg > 0d)
                {
                    result.Add(new MineResourceEntry(CompactText(matcher.Groups[1].Value), kg));
                }

                matcher = matcher.NextMatch();
            }

            return result;
        }

        private static void PublishMineChatReport(string address, string html, MineDiggReport report, bool isNewReport)
        {
            if (report != null && !isNewReport)
                return;

            var message = string.Empty;
            if (report != null)
            {
                message = BuildDiggChatMessage(report);
            }
            else if (html.IndexOf("Обнаружены ресурсы", StringComparison.OrdinalIgnoreCase) >= 0)
            {
                message = CompactFromMarker(html, "Обнаружены ресурсы");
            }
            else if (html.IndexOf("Добыты ресурсы", StringComparison.OrdinalIgnoreCase) >= 0)
            {
                message = CompactFromMarker(html, "Добыты ресурсы");
            }
            else if (html.IndexOf("Вы не нашли ни одного ресурса", StringComparison.OrdinalIgnoreCase) >= 0)
            {
                message = "Вы не нашли ни одного ресурса";
            }

            if (string.IsNullOrEmpty(message) || AppVars.MainForm == null)
                return;

            AppVars.MainForm.WriteChatMsgSafe("<font color=#333399><b>[auto_mine/mine_ajax]</b></font> " + EscapeHtml(message));
            AppLog.i(AutoMineRuntime.TraceChain, "MineAjaxPhp", "chat report posted: address=" + (address ?? string.Empty) + ", message=" + message);
        }

        private static string BuildDiggChatMessage(MineDiggReport report)
        {
            var builder = new StringBuilder();
            builder.Append("Ресурсы: ");
            if (report.Resources.Count == 0)
            {
                builder.Append("не распознаны");
            }
            else
            {
                for (var i = 0; i < report.Resources.Count; i++)
                {
                    if (i > 0)
                        builder.Append(", ");
                    builder.Append(report.Resources[i].Name);
                    builder.Append(" (");
                    builder.Append(FormatKg(report.Resources[i].Kg));
                    builder.Append(" кг)");
                }
            }

            if (report.ServerDelaySeconds > 0)
            {
                builder.Append(". Задержка: ");
                builder.Append((report.ServerDelaySeconds + 2).ToString(CultureInfo.InvariantCulture));
                builder.Append(" сек.");
            }

            return builder.ToString();
        }

        private static string CompactFromMarker(string value, string marker)
        {
            var index = value.IndexOf(marker, StringComparison.OrdinalIgnoreCase);
            return CompactText(index >= 0 ? value.Substring(index) : value);
        }

        private static string CompactText(string value)
        {
            if (string.IsNullOrEmpty(value))
                return string.Empty;

            var result = Regex.Replace(value.Replace('\u00A0', ' '), "<[^>]+>", " ");
            result = Regex.Replace(result, "\\s+", " ").Trim();
            return result.Length > 400 ? result.Substring(0, 400) + "..." : result;
        }

        private static string FormatKg(double value)
        {
            var formatted = value.ToString("0.##", CultureInfo.InvariantCulture);
            return formatted;
        }

        private static string EscapeHtml(string value)
        {
            if (string.IsNullOrEmpty(value))
                return string.Empty;

            return value.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("\"", "&quot;").Replace("'", "&#39;");
        }

        private sealed class MineDiggReport
        {
            internal readonly string EventKey;
            internal readonly string ReportText;
            internal readonly int ServerDelaySeconds;
            internal readonly List<MineResourceEntry> Resources;
            internal readonly Dictionary<string, double> DeltaByResourceKg;

            internal MineDiggReport(string eventKey, string reportText, int serverDelaySeconds, List<MineResourceEntry> resources, Dictionary<string, double> deltaByResourceKg)
            {
                EventKey = eventKey;
                ReportText = reportText;
                ServerDelaySeconds = serverDelaySeconds;
                Resources = resources ?? new List<MineResourceEntry>();
                DeltaByResourceKg = deltaByResourceKg ?? new Dictionary<string, double>(StringComparer.OrdinalIgnoreCase);
            }
        }

        private sealed class MineResourceEntry
        {
            internal readonly string Name;
            internal readonly double Kg;

            internal MineResourceEntry(string name, double kg)
            {
                Name = name;
                Kg = kg;
            }
        }
    }
}
